### PR TITLE
Add ECG demo (cardiac coherence using LF/HF) 

### PR DESCRIPTION
**Pull request by bertrandlalo**

### CODE REVIEW

File: coherence/main.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        topics: [raw, filtered, features, rr, scaled_rr ]
+
+    # Define apps for feedback
+    # -----------------------
+    - id: ui
+      module: timeflux_ui.nodes.ui
+      class: UI
+      params:
+        routes:
+          ecg_cardiac: www  # now open http://localhost:8000/ecg_cardiac/
```
Comment (mesca) [inline coherence/main.yaml]: Rename route to `coherence`

File: coherence/www/assets/app.css [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+html, body, #container {
+    height: 100%;
+    padding: 0;
+}
+
+#circle-container {
+    display: flex;
+    align-items: center;
+    justify-content: center;
+    height: 500pt;  /* todo: unhardcode this height */
```
Comment (mesca) [inline coherence/www/assets/app.css]: Indeed ;)

File: coherence/www/assets/app.css [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+#circle {
+    background-color: var(--circle-color);
+    opacity: var(--circle-opacity);
+    width: var(--circle-width);
+    height: var(--circle-height);
+    border-radius: 50%;
+    /*#transition: background-color 2s ease-in-out;*/
```
Comment (mesca) [inline coherence/www/assets/app.css]: Remove

File: coherence/www/assets/app.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+ * @param {string|number} value
+ */
+function set_css_var(name, value) {
+  document.documentElement.style.setProperty(name, value);
+}
+
+/**
+ * Estimate min/max of circle giveb container size
+ */
+function resize() {
```
Comment (mesca) [inline coherence/www/assets/app.js:130]: This should be achievable with CSS only.

File: coherence/README.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ <img src="img/montage_ecg.png" width="40%">
+ 
+/!\ If you revert the IN+ (red) and IN-(black) electrodes, the ECG QRS will be inverted and the cardiac peaks won't be estimated correctly.   
+
+
+## Demo 
+
+```bash
+	conda activate timeflux-hackathon
+	timeflux -d graphs/ecg_cardiac.yaml
```
Comment (mesca) [inline coherence/README.md]: Update graph name

File: coherence/README.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# Cardiac (ECG) 
+In this demo, we acquire heart signal using ECG sensor, we detect heart beats, estimate HRV in realtime and then HRV spectrum, to compute "coherence" biomarkers that is the ratio between low frequency and high frequency. The feedback is a circle which radius depends on heart rate and the color depends on coherence level. The idea is to be able to control this circle.  
+
+## Sensor montage
+You need to plug the ECG at inpuut A1 of Bitalino and place the electrode as follow: 
+
+ <img src="img/montage_ecg.png" width="40%">
+ 
+/!\ If you revert the IN+ (red) and IN-(black) electrodes, the ECG QRS will be inverted and the cardiac peaks won't be estimated correctly.
```
Comment (mesca) [inline coherence/README.md]: Use emoji instead: :warning:

File: coherence/www/assets/app.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+let io = new IO();
+
+// Prepare elements from DOM
+let _td_features = {};
+const columns = ["bpm", "lf", "hf", "lf/hf"]; // columns of EMG activation bursts
+columns.forEach(function (column) {
+  _td_features[column] = document.getElementById(column);
+});
+const td_bpm = document.getElementById("bpm");
+const _gradientColors = [
```
Comment (mesca) [inline coherence/www/assets/app.js:12]: Would be great to report the colors in the README or at least to provide a succinct explanation

**Review by mesca — CHANGES_REQUESTED** [review]
The cardiac features are lagging behind in the UI. Is this because of the filtering. Because of the graph rate?

File: coherence/main.yaml [hunk 8]
```diff
@@ -0,0 +1,54 @@
+import:
+# input graphs (choose one amongst the three! )
+# ----------------------------
+  - graphs/bitalino.yaml # uncomment this line to run with bitalino device
+  # - graphs/replay.yaml # uncomment this line to run on replayed data
```
Comment (mesca) [inline coherence/main.yaml:5]: Default should be sine so one can launch the demo without any hardware

File: coherence/www/assets/app.js [hunk 9]
```diff
@@ -0,0 +1,139 @@
+"use strict";
+
+let io = new IO();
+
+// Prepare elements from DOM
+let _td_features = {};
+const columns = ["bpm", "lf", "hf", "lf/hf"]; // columns of EMG activation bursts
```
Comment (mesca) [inline coherence/www/assets/app.js]: EMG ?
;)

File: coherence/www/assets/app.js [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+// Scaled RR interval defines the circle radius
+io.on("radius", (data) => {
+  if (Object.keys(data).length > 0) {
+    let row = data[Object.keys(data)[Object.keys(data).length - 1]]; // Last row
+    let column = Object.keys(row)[0]; // First column
+    let value = row[column]; // Value
+    value = Math.tanh(value); // Ensure value between 0 and 1
+    value = 1 - Math.tanh(value); // circle radius should decrease when interval increase (ie. when BPM decrease)
+    const px = value * (max - min) + min + "px"; // Transform to pixels
```
Comment (mesca) [inline coherence/www/assets/app.js:63]: Use percentage?

**Review by mesca — CHANGES_REQUESTED** [review]

File: coherence/www/assets/app.js [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    max = Math.max.apply(Math, [a, b]);
+  return inclusive ? this >= min && this <= max : this > min && this < max;
+};
+
+/**
+ * Return color gradient
+ *
+ * @param {number}  x variable in [0, 1]
+ */
+const color_gradient = (x) => {
```
Comment (mesca) [inline coherence/www/assets/app.js]: Harmonize function declaration style

File: coherence/main.yaml [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    # Define apps for feedback
+    # -----------------------
+    - id: ui
+      module: timeflux_ui.nodes.ui
+      class: UI
+      params:
+        routes:
+          ecg_cardiac: www  # now open http://localhost:8000/ecg_cardiac/
+        settings:
+          circle:
```
Comment (mesca) [inline coherence/main.yaml:39]: Use the route name

File: coherence/nodes/cardiac.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
+from timeflux.core.node import Node
+import numpy as np
+import pandas as pd
+
+
+class CardiacFreqMarkers(Node):
```
Comment (mesca) [inline coherence/nodes/cardiac.py:6]: Comment inputs and output

File: coherence/nodes/events.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
+import pandas as pd
+
+from timeflux.core.node import Node
+
+
+class ToSignal(Node):
```
Comment (mesca) [inline coherence/nodes/events.py]: Give a more explicit name? `EventToSignal`?

File: coherence/nodes/events.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    """ Serialize one column
+    Attributes:
```
Comment (mesca) [inline coherence/nodes/events.py:9]: Add line breaks to make it more readable

File: coherence/nodes/events.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    >>> node.update()
+    >>> node.o.data
+                                label  a     b
+        2000-01-01 00:00:00.001   bar  2  20.0
+        2000-01-01 00:00:00.000   foo  1  10.0
+        2000-01-01 00:00:00.004   foo  1   NaN
+    """
+
+    def __init__(self, meta_keys, labels=None, drop_label=True):
+        super().__init__()
```
Comment (mesca) [inline coherence/nodes/events.py]: Why this line?

**Review by mesca — CHANGES_REQUESTED** [review]

File: coherence/www/assets/app.js (hunk 4 above)
Comment (bertrandlalo) [reply coherence/www/assets/app.js:130]: Tell me how 🙏😇

File: coherence/www/assets/app.js (hunk 10 above)
Comment (bertrandlalo) [reply coherence/www/assets/app.js:63]: ?

File: coherence/www/assets/app.js (hunk 4 above)
Comment (mesca) [reply coherence/www/assets/app.js:130]: You could probably use CSS percentage and simplify the code.

File: coherence/www/assets/app.js (hunk 7 above)
Comment (mesca) [reply coherence/www/assets/app.js:12]: Great. I would avoid mentioning Gandhi though (here and in the README).

**Pull request merged**